src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/ObjectHeader.cs

```diff
@@ -183,32 +186,302 @@ public static unsafe void SetSyncEntryIndex(int* pHeader, int syncIndex)
             }
             while (Interlocked.CompareExchange(ref *pHeader, newBits, oldBits) != oldBits);
         }
+
+        //
+        // A few words about spinning choices:
+        //
+        // Most locks are not contentious. In fact most locks provide exclusive access safety, but in reality are used by
+        // one thread. And then there are locks that do see multiple threads, but the accesses are short and not overlapping.
+        // Thin lock is an optimization for such scenarios.
+        //
+        // If we see a thin lock held by other thread for longer than ~5 microseconds, we will "inflate" the lock
+        // and let the adaptive spinning in the fat Lock sort it out whether we have a contentious lock or a long-held lock.
+        //
+        // Another thing to consider is that SpinWait(1) is calibrated to about 35-50 nanoseconds.
+        // It can take much longer only if nop/pause takes much longer, which it should not, as that would be getting
+        // close to the RAM latency.
+        //
+        // Considering that taking and releasing the lock takes 2 CAS instructions + some overhead, we can estimate shortest
+        // time the lock can be held to be in hundreds of nanoseconds. Thus it is unlikely to see more than
+        // 8-10 threads contending for the lock without inflating it. Therefore we can expect to acquire a thin lock in
+        // under 16 tries.
+        //
+        // As for the backoff strategy we have two choices:
+        // Exponential back-off with a lmit:
+        //   0, 1, 2, 4, 8, 8, 8, 8, 8, 8, 8, . . . .
+        //
+        // Linear back-off
+        //   0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, . . . .
+        //
+        // In this case these strategies are close in terms of average and worst case latency, so we will prefer linear
+        // back-off as it favors micro-contention scenario, which we expect.
+        //
+
+        // Returs:
+        //   1 - success
+        //   0 - failed
+        //   syncIndex - retry with the Lock
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe int Acquire(object obj)
+        {
+            // Lock.s_processorCount is lazy-initialized at fist contention.
+            // untill then assume multicore
+            return TryAcquire(obj, oneShot: false);
+        }
+
+        // 1 - success
+        // 0 - failed
+        // syncIndex - retry with the Lock
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe int TryAcquire(object obj, bool oneShot = true)
+        {
+            ArgumentNullException.ThrowIfNull(obj);
+
+            Debug.Assert(!(obj is Lock),
+                "Do not use Monitor.Enter or TryEnter on a Lock instance; use Lock methods directly instead.");
+
+            int currentThreadID = Environment.CurrentManagedThreadIdUnchecked;
+            // if thread ID is uninitialized too big, we do uncommon part.
```

Comment:
```suggestion
            Debug.Assert((uint)ManagedThreadId.None > (uint)SBLK_MASK_LOCK_THREADID);
            // if thread ID is uninitialized too big, we do uncommon part.
```
To make it obvious that we are depending on ManagedThreadId implementation details here.